src/reducers/index.ts

```diff
@@ -18,9 +42,64 @@ function reducer(state: Readonly<StateBase>, action: Action): StateBase {
   };
 }
 
+/**
+ * Exclude these actions from the history completely.
+ */
+const excludeActionsInHistory = [DATASET_INLINE_RECEIVE,
+  RESULT_RECEIVE,
+  RESULT_REQUEST,
+  SET_CONFIG,
+];
+
+/**
+ * A list of actions that can be initiated by a user.
+ *
+ * Each of these will start a new 'undo group'. Non-user actions will be put into the group
+ * of the preceding user action if one is available. If none is available it will be put
+ * into its own group.
+ */
+const userActions = [
+  // Dataset Actions
+  DATASET_URL_REQUEST,
+  DATASET_URL_RECEIVE,
+  DATASET_INLINE_RECEIVE,
+  // Shelf Actions,
+  SHELF_CLEAR,
+  SHELF_MARK_CHANGE_TYPE,
+  SHELF_FIELD_ADD,
+  SHELF_FIELD_AUTO_ADD,
+  SHELF_FIELD_REMOVE,
+  SHELF_FIELD_MOVE,
+  SHELF_FUNCTION_CHANGE,
+  SHELF_SPEC_LOAD,
+  SHELF_SPEC_PREVIEW,
+  SHELF_SPEC_PREVIEW_DISABLE,
+];
+
+let _groupId = 0;
+function getNextGroupId(): number {
+  _groupId += 1;
+  return _groupId;
+}
+
+const groupAction = (action: Action, currentState: State, previousHistory: StateWithHistory<State>): any => {
+  const currentActionType = action.type;
+
+  if (userActions.indexOf(currentActionType) >= 0) {
```

Comment:
true i was also thinking of using Array.includes or an ES6 set if typescript supports that. I will look into it! I'm probably leaning towards an ES6 set